security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -0,0 +1,334 @@
+				// could connect again with updated token.
+				// When nodeagent stops stream by sending envoy error response, it's Ok not to remove secret
+				// from secret cache because cache has auto-evication.
+				log.Debugf("Close connection with %q", con.proxyID)
+				return fmt.Errorf("streaming connection with %q closed", con.proxyID)
+			}
+
+			if err := pushSDS(con); err != nil {
+				log.Errorf("SDS failed to push: %v", err)
+				return err
+			}
+		}
+	}
+}
+
+func (s *sdsservice) FetchSecrets(ctx context.Context, discReq *xdsapi.DiscoveryRequest) (*xdsapi.DiscoveryResponse, error) {
+	token, err := getCredentialToken(ctx)
+	if err != nil {
+		log.Errorf("Failed to get credential token: %v", err)
+		return nil, err
+	}
+
+	spiffeID, err := parseDiscoveryRequest(discReq)
+	if err != nil {
+		log.Errorf("Failed to parse discovery request: %v", err)
+		return nil, err
+	}
+
+	secret, err := s.st.GenerateSecret(ctx, discReq.Node.Id, spiffeID, token)
+	if err != nil {
+		log.Errorf("Failed to get secret for proxy %q from secret cache: %v", discReq.Node.Id, err)
+		return nil, err
+	}
+
+	return sdsDiscoveryResponse(secret, discReq.Node.Id)
+}
+
+// NotifyProxy send notification to proxy about secret update,
+// SDS will close streaming connection is secret is nil.
+func NotifyProxy(proxyID string, secret *model.SecretItem) error {
+	conn := sdsClients[proxyID]
+	if conn == nil {
+		log.Errorf("No connection with id %q can be found", proxyID)
+		return fmt.Errorf("no connection with id %q can be found", proxyID)
+	}
+	conn.secret = secret
+
+	conn.pushChannel <- &sdsEvent{}
+	return nil
+}
+
+func parseDiscoveryRequest(discReq *xdsapi.DiscoveryRequest) (string /*spiffeID*/, error) {
+	if discReq.Node.Id == "" {
+		return "", fmt.Errorf("discovery request %+v missing node id", discReq)
+	}
+
+	if len(discReq.ResourceNames) != 1 || !strings.HasPrefix(discReq.ResourceNames[0], util.URIScheme) {
+		return "", fmt.Errorf("discovery request %+v has invalid resourceNames %+v", discReq, discReq.ResourceNames)
+	}
+
+	return discReq.ResourceNames[0], nil
+}
+
+func getCredentialToken(ctx context.Context) (string, error) {
+	metadata, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return "", fmt.Errorf("unable to get metadata from incoming context")
+	}
+
+	if h, ok := metadata[CredentialTokenHeaderKey]; ok {
+		if len(h) != 1 {
+			return "", fmt.Errorf("credential token must have 1 value in gRPC metadata but got %d", len(h))
+		}
+		return h[0], nil
+	}
+
+	return "", fmt.Errorf("no credential token is found")
+}
+
+func addConn(proxyID string, conn *sdsConnection) {
+	sdsClientsMutex.Lock()
+	defer sdsClientsMutex.Unlock()
+	sdsClients[proxyID] = conn
+}
+
+func removeConn(proxyID string) {
+	sdsClientsMutex.Lock()
+	defer sdsClientsMutex.Unlock()
+	delete(sdsClients, proxyID)
+}
+
+func pushSDS(con *sdsConnection) error {
+	log.Infof("SDS: push for proxy:%q", con.proxyID)
```

Comment:
"SDS: push for proxy" is not quite clear, maybe something like

"SDS: push from node agent to proxy:"

Reply:
updated